CONFIGURATION.md
# Configuration for Gaurun

The configuration file format for Gaurun is [TOML](https://github.com/toml-lang/toml).

The configuration for Gaurun has some sections. The example is [here](conf/gaurun.toml).

 * [Core Section](#core-section)
 * [iOS Section](#ios-section)
 * [Android Section](#android-section)
 * [Log Section](#log-section)

## Core Section

| name             | type   | description                                                                     | default          | note                                                                         |
| ---------------- | ------ | ------------------------------------------------------------------------------- | ---------------- | ---------------------------------------------------------------------------- |
| port             | string | port number or unix socket path                                                 | 1056             | e.g.)1056, unix:/tmp/gaurun.sock <br/> `-p` option can overwrite             |
| workers          | int64  | number of workers for push notification                                         | runtime.NumCPU() | `-w` options can overwrite                                                   |
| queues           | int64  | size of internal queue for push notification                                    | 8192             | `-q` options can overwrite                                                   |
| notification_max | int64  | limit of push notifications once                                                | 100              |                                                                              |
| pusher_max       | int64  | maximum goroutines for asynchronous pushing                                     | 0                | If the value is less than or equal to zero, each worker pushes synchronously |
| shutdown_timeout | int64  | timeout to wait for connections to return to idle when server shutdown (second) | 10               |                                                                              |
| pid              | string | path to pid file                                                                |                  |                                                                              |

## iOS Section

| name                | type   | description                                              | default          | note |
| ------------------- | ------ | -------------------------------------------------------- | ---------------- | ---- |
| enabled             | bool   | On/Off for push notication to APNs                       | true             |      |
| pem_cert_path       | string | certification file path for APNs                         |                  |      |
| pem_key_path        | string | secret key file path for APNs                            |                  |      |
| pem_key_passphrase  | string | secret key file pass phrase for APNs                     |                  |      |
| token_auth_key_path | string | secret APNs auth key file (.p8) for token based provider |                  |      |
| token_auth_key_id   | string | APNs key id for token based provider                     |                  |      |
| token_auth_team_id  | string | APNs team id for token based provider                    |                  |      |
| sandbox             | bool   | On/Off for sandbox environment                           | true             |      |
| retry_max           | int    | maximum retry count for push notication to APNs          | 1                |      |
| timeout             | int    | timeout for push notification to APNs                    | 5                |      |
| keepalive_timeout   | int    | time for continuing keep-alive connection to APNs        | 90               |      |
| keepalive_conns     | int    | number of keep-alive connection to APNs                  | runtime.NumCPU() |      |
| topic               | string | the assigned value of `apns-topic` for Request headers   |                  |      |

`topic` is mandatory when the client is connected using the certificate that supports multiple topics.

## Android Section

| name              | type   | description                                      | default          | note |
| ----------------- | ------ | ------------------------------------------------ | ---------------- | ---- |
| enabled           | bool   | On/Off for push notication to FCM                | true             |      |
| apikey            | string | API key string for FCM                           |                  |      |
| timeout           | int    | timeout for push notication to FCM               | 5(sec)           |      |
| keepalive_timeout | int    | time for continuing keep-alive connection to FCM | 90               |      |
| keepalive_conns   | int    | number of keep-alive connection to FCM           | runtime.NumCPU() |      |
| retry_max         | int    | maximum retry count for push notication to FCM   | 1                |      |

## Log Section

| name       | type   | description     | default | note                              |
| ---------- | ------ | --------------- | ------- | --------------------------------- |
| access_log | string | access log path | stdout  |                                   |
| error_log  | string | error log path  | stderr  |                                   |
| level      | string | log level       | error   | panic,fatal,error,warn,info,debug |

`access_log` and `error_log` are allowed to give not only file-path but `stdout` and `stderr` and `discard`.


gaurun/conf.go
package gaurun

import (
	"net/http"
	"net/url"
	"runtime"
	"strconv"
	"sync/atomic"

	"github.com/BurntSushi/toml"
)

type ConfToml struct {
	Core    SectionCore    `toml:"core"`
	Android SectionAndroid `toml:"android"`
	Ios     SectionIos     `toml:"ios"`
	Log     SectionLog     `toml:"log"`
}

type SectionCore struct {
	Port               string `toml:"port"`
	WorkerNum          int64  `toml:"workers"`
	QueueNum           int64  `toml:"queues"`
	NotificationMax    int64  `toml:"notification_max"`
	PusherMax          int64  `toml:"pusher_max"`
	ShutdownTimeout    int64  `toml:"shutdown_timeout"`
	Pid                string `toml:"pid"`
	AllowsEmptyMessage bool   `toml:"allows_empty_message"`
}

type SectionAndroid struct {
	Enabled          bool   `toml:"enabled"`
	ApiKey           string `toml:"apikey"`
	Timeout          int    `toml:"timeout"`
	KeepAliveTimeout int    `toml:"keepalive_timeout"`
	KeepAliveConns   int    `toml:"keepalive_conns"`
	RetryMax         int    `toml:"retry_max"`
}

type SectionIos struct {
	Enabled          bool   `toml:"enabled"`
	PemCertPath      string `toml:"pem_cert_path"`
	PemKeyPath       string `toml:"pem_key_path"`
	PemKeyPassphrase string `toml:"pem_key_passphrase"`
	TokenAuthKeyPath string `toml:"token_auth_key_path"`
	TokenAuthKeyID   string `toml:"token_auth_key_id"`
	TokenAuthTeamID  string `toml:"token_auth_team_id"`
	Sandbox          bool   `toml:"sandbox"`
	RetryMax         int    `toml:"retry_max"`
	Timeout          int    `toml:"timeout"`
	KeepAliveTimeout int    `toml:"keepalive_timeout"`
	KeepAliveConns   int    `toml:"keepalive_conns"`
	Topic            string `toml:"topic"`
}

type SectionLog struct {
	AccessLog string `toml:"access_log"`
	ErrorLog  string `toml:"error_log"`
	Level     string `toml:"level"`
}

func BuildDefaultConf() ConfToml {
	numCPU := runtime.NumCPU()

	var conf ConfToml
	// Core
	conf.Core.Port = "1056"
	conf.Core.WorkerNum = int64(numCPU)
	conf.Core.QueueNum = 8192
	conf.Core.NotificationMax = 100
	conf.Core.PusherMax = 0
	conf.Core.ShutdownTimeout = 10
	conf.Core.Pid = ""
	conf.Core.AllowsEmptyMessage = false
	// Android
	conf.Android.ApiKey = ""
	conf.Android.Enabled = true
	conf.Android.Timeout = 5
	conf.Android.KeepAliveTimeout = 90
	conf.Android.KeepAliveConns = numCPU
	conf.Android.RetryMax = 1
	// iOS
	conf.Ios.Enabled = true
	conf.Ios.PemCertPath = ""
	conf.Ios.PemKeyPath = ""
	conf.Ios.TokenAuthKeyPath = ""
	conf.Ios.TokenAuthKeyID = ""
	conf.Ios.TokenAuthTeamID = ""
	conf.Ios.Sandbox = true
	conf.Ios.RetryMax = 1
	conf.Ios.Timeout = 5
	conf.Ios.KeepAliveTimeout = 90
	conf.Ios.KeepAliveConns = numCPU
	conf.Ios.Topic = ""
	// log
	conf.Log.AccessLog = "stdout"
	conf.Log.ErrorLog = "stderr"
	conf.Log.Level = "error"
	return conf
}

func LoadConf(confGaurun ConfToml, confPath string) (ConfToml, error) {
	_, err := toml.DecodeFile(confPath, &confGaurun)
	if err != nil {
		return confGaurun, err
	}
	return confGaurun, nil
}

func ConfigPushersHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != "PUT" {
		sendResponse(w, "method must be PUT", http.StatusBadRequest)
		return
	}

	values, err := url.ParseQuery(r.URL.RawQuery)
	if err != nil {
		LogError.Error(err.Error())
		sendResponse(w, "url parameters could not be parsed", http.StatusBadRequest)
		return
	}

	in := ""
	for k, v := range values {
		if k == "max" {
			in = v[0]
			break
		}
	}

	if in == "" {
		sendResponse(w, "malformed value", http.StatusBadRequest)
		return
	}

	newPusherMax, err := strconv.ParseInt(in, 0, 64)
	if err != nil {
		LogError.Error(err.Error())
		sendResponse(w, "malformed value", http.StatusBadRequest)
		return
	}

	if newPusherMax < 0 {
		sendResponse(w, "malformed value", http.StatusBadRequest)
		return
	}

	atomic.StoreInt64(&ConfGaurun.Core.PusherMax, newPusherMax)

	sendResponse(w, "ok", http.StatusOK)
}
